hax/hax/hax.py

```diff
@@ -100,15 +100,16 @@ def main():
 
     ffi = HaxFFI()
     herald = DeliveryHerald()
-    motr = Motr(queue=q,
-                ffi=ffi,
-                herald=herald,
-                consul_util=util)
+    motr = Motr(planner=planner, ffi=ffi, herald=herald, consul_util=util)
 
     # Note that consumer thread must be started before we invoke motr.start(..)
     # Reason: hax process will send entrypoint request and somebody needs
     # to reply it.
-    consumer = _run_qconsumer_thread(q, motr, herald, util)
+
+    # TODO make the number of threads configurable
+    consumer_threads = [
```

Comment:
I think we can still configure number of consumer threads instead of a constant, anyway execution of threads is restricted to the number of cpus, so how about creating threads equal to `os.cpu_count()`?

Reply:
But we should make sure that the thread execution does not block.

Reply:
As far as I remember, the formula usually looks like this:
`Number of threads = Number of Available Cores / (1 - Blocking Coefficient)` ([source](https://superuser.com/questions/664550/optimum-number-of-threads-while-multitasking))

So it really depends how much of time the threads can do something meaningful and how much of time they will be waiting for an IO operation. In other words, we probably need to experiment and propose a rule of a thumb for this setting.

Reply:
In worst case every thread may block (although the best efforts should be to avoid executing a blocking operation), in such cases if a work item is going to block a thread, an additional handler thread must be created which must be destroyed again after completion of the blocking operation via some routines, e.g. block_enter(), block_leave(). But this can complicate the design, thus for now we can come up with a formula to derive number of threads and try to avoid blocking the consumer thread.

Reply:
Even adding a new thread like you described can decrease the performance (in case when maintaining a thread, context switching and GIL acquiring brings more problems than solves anything).

When threads block on IO the most of time, then obviously number of threads should be `round(N * os.cpu_count())` where N > 1. When the load of Motr-related messages is not that high, then too much of parallelism doesn't make sense (even on a machine with 32 CPU cores). That brings us to `round(M * os.cpu_count())` where 0 < M <= 1.

I think that we need to gather some statistics (like mean or max time that commands spend in the backlog) and warn if the values are suspiciously big (i.e. processing is too slow). That can helps us to held the experiments and thus calibrate the recommended coefficients.